File: src/ingestion/chains/studenac.ts
/**
 * Studenac Chain Adapter
 *
 * Adapter for parsing Studenac retail chain price data files.
 * Studenac uses XML format with store information embedded in the XML structure.
 * Store resolution is based on portal ID within the XML content.
 */

import type { DiscoveredFile } from '../core/types'
import type { XmlFieldMapping } from '../parsers/xml'
import { BaseXmlAdapter } from './base'
import { CHAIN_CONFIGS } from './config'

/**
 * XML field mapping for Studenac files.
 * Maps Studenac's XML elements to NormalizedRow fields.
 * Studenac XML structure typically has store info and product data nested.
 */
const STUDENAC_FIELD_MAPPING: XmlFieldMapping = {
  storeIdentifier: (item) => {
    // Extract store ID from item or parent store block
    const storeId =
      (item['store_id'] as string) ??
      (item['storeId'] as string) ??
      (item['Store'] as Record<string, unknown>)?.['Id'] as string ??
      null
    return storeId ? String(storeId) : null
  },
  externalId: 'code',
  name: 'name',
  description: 'description',
  category: 'category',
  subcategory: 'subcategory',
  brand: 'brand',
  unit: 'unit',
  unitQuantity: 'quantity',
  price: 'price',
  discountPrice: 'discount_price',
  discountStart: 'discount_start',
  discountEnd: 'discount_end',
  barcodes: 'barcode',
  imageUrl: 'image_url',
  // Croatian price transparency fields
  unitPrice: 'unit_price',
  unitPriceBaseQuantity: 'unit_price_quantity',
  unitPriceBaseUnit: 'unit_price_unit',
  lowestPrice30d: 'lowest_price_30d',
  anchorPrice: 'anchor_price',
  anchorPriceAsOf: 'anchor_price_date',
}

/**
 * Alternative field mapping for Studenac XML files (uppercase/different naming).
 */
const STUDENAC_FIELD_MAPPING_ALT: XmlFieldMapping = {
  storeIdentifier: (item) => {
    const storeId =
      (item['StoreId'] as string) ??
      (item['STORE_ID'] as string) ??
      (item['Poslovnica'] as Record<string, unknown>)?.['Id'] as string ??
      null
    return storeId ? String(storeId) : null
  },
  externalId: 'Sifra',
  name: 'Naziv',
  description: 'Opis',
  category: 'Kategorija',
  subcategory: 'Podkategorija',
  brand: 'Marka',
  unit: 'Jedinica',
  unitQuantity: 'Kolicina',
  price: 'Cijena',
  discountPrice: 'AkcijskaCijena',
  discountStart: 'PocetakAkcije',
  discountEnd: 'KrajAkcije',
  barcodes: 'Barkod',
  imageUrl: 'Slika',
  // Croatian price transparency fields
  unitPrice: 'CijenaZaJedinicuMjere',
  unitPriceBaseQuantity: 'JedinicaMjereKolicina',
  unitPriceBaseUnit: 'JedinicaMjereOznaka',
  lowestPrice30d: 'NajnizaCijena30Dana',
  anchorPrice: 'SidrenaCijena',
  anchorPriceAsOf: 'SidrenaCijenaDatum',
}

/**
 * Studenac chain adapter implementation.
 * Extends BaseXmlAdapter for common XML parsing functionality.
 *
 * Studenac portal: https://www.studenac.hr/popis-maloprodajnih-cijena
 * Supports date-based discovery via setDiscoveryDate() method.
 */
export class StudenacAdapter extends BaseXmlAdapter {
  /** Date to discover files for (YYYY-MM-DD format, set by CLI before discovery) */
  private discoveryDate: string | null = null

  constructor() {
    super({
      slug: 'studenac',
      name: 'Studenac',
      supportedTypes: ['xml'],
      chainConfig: CHAIN_CONFIGS.studenac,
      fieldMapping: STUDENAC_FIELD_MAPPING,
      alternativeFieldMapping: STUDENAC_FIELD_MAPPING_ALT,
      defaultItemsPath: 'products.product',
      filenamePrefixPatterns: [
        /^Studenac[_-]?/i,
        /^cjenik[_-]?/i,
        /^SUPERMARKET[_-]?/i,
      ],
      rateLimitConfig: {
        requestsPerSecond: 2,
        maxRetries: 3,
      },
    })
  }

  /**
   * Set the date to use for discovery.
   * @param date - Date in YYYY-MM-DD format
   */
  setDiscoveryDate(date: string): void {
    this.discoveryDate = date
  }

  /**
   * Extract date from Studenac filename.
   * Studenac filenames often contain dates in format: YYYY-MM-DD or DD-MM-YYYY
   * Example: SUPERMARKET-Bijela_uvala_5_FUNTANA-T598-229-2026-12-29-07-00-14-559375.xml
   */
  private extractDateFromFilename(filename: string): string | null {
    // Try YYYY-MM-DD pattern
    const isoMatch = filename.match(/(\d{4})-(\d{2})-(\d{2})/)
    if (isoMatch) {
      return `${isoMatch[1]}-${isoMatch[2]}-${isoMatch[3]}`
    }

    // Try DD-MM-YYYY pattern
    const euMatch = filename.match(/(\d{2})-(\d{2})-(\d{4})/)
    if (euMatch) {
      return `${euMatch[3]}-${euMatch[2]}-${euMatch[1]}`
    }

    return null
  }

  /**
   * Discover available Studenac price files.
   *
   * Studenac's portal structure may vary. This implementation:
   * - Fetches the portal HTML
   * - Extracts XML file links
   * - Filters by date if setDiscoveryDate was called
   *
   * @returns Array of discovered files (filtered by date if setDiscoveryDate was called)
   */
  async discover(): Promise<DiscoveredFile[]> {
    const discoveredFiles: DiscoveredFile[] = []
    const seenUrls = new Set<string>()

    console.log(`[DEBUG] Fetching Studenac portal: ${this.config.baseUrl}`)

    try {
      const response = await fetch(this.config.baseUrl, {
        headers: {
          'User-Agent': 'Mozilla/5.0 (compatible; PriceTracker/1.0)',
          'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        },
      })

      if (!response.ok) {
        console.error(`Failed to fetch Studenac portal: ${response.status} ${response.statusText}`)
        console.error(`  URL: ${this.config.baseUrl}`)
        return []
      }

      const html = await response.text()

      // Extract XML file links
      const xmlPattern = /href=["']([^"']*\.xml(?:\?[^"']*)?)["']/gi

      let match: RegExpExecArray | null
      while ((match = xmlPattern.exec(html)) !== null) {
        const href = match[1]
        const fileUrl = href.startsWith('http') ? href : new URL(href, this.config.baseUrl).toString()

        // Skip duplicates
        if (seenUrls.has(fileUrl)) {
          continue
        }
        seenUrls.add(fileUrl)

        // Extract filename from URL
        const filename = this.extractFilenameFromUrl(fileUrl)
        const fileDate = this.extractDateFromFilename(filename)

        // Filter by date if discoveryDate is set
        if (this.discoveryDate && fileDate && fileDate !== this.discoveryDate) {
          continue
        }

        discoveredFiles.push({
          url: fileUrl,
          filename,
          type: 'xml',
          size: null,
          lastModified: fileDate ? new Date(fileDate) : null,
          metadata: {
            source: 'studenac_portal',
            discoveredAt: new Date().toISOString(),
            ...(fileDate && { portalDate: fileDate }),
          },
        })
      }

      return discoveredFiles
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error)
      console.error(`Error discovering Studenac files: ${errorMessage}`)
      console.error(`  URL: ${this.config.baseUrl}`)
      return []
    }
  }
}

/**
 * Create a Studenac adapter instance.
 */
export function createStudenacAdapter(): StudenacAdapter {
  return new StudenacAdapter()
}

File: services/price-service/internal/adapters/chains/studenac.go
package chains

import (
	"fmt"
	"io"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/kosarica/price-service/internal/adapters/base"
	"github.com/kosarica/price-service/internal/adapters/config"
	"github.com/kosarica/price-service/internal/parsers/xml"
	"github.com/kosarica/price-service/internal/types"
)

// studenacFieldMapping is the primary field mapping for Studenac XML files (lowercase/snake_case)
var studenacFieldMapping = xml.XmlFieldMapping{
	// StoreIdentifier uses an extractor function - we'll set it at runtime
	ExternalID:            types.StringPtr("code"),
	Name:                  "name",
	Description:           types.StringPtr("description"),
	Category:              types.StringPtr("category"),
	Subcategory:           types.StringPtr("subcategory"),
	Brand:                 types.StringPtr("brand"),
	Unit:                  types.StringPtr("unit"),
	UnitQuantity:          types.StringPtr("quantity"),
	Price:                 "price",
	DiscountPrice:         types.StringPtr("discount_price"),
	DiscountStart:         types.StringPtr("discount_start"),
	DiscountEnd:           types.StringPtr("discount_end"),
	Barcodes:              types.StringPtr("barcode"),
	ImageURL:              types.StringPtr("image_url"),
	UnitPrice:             types.StringPtr("unit_price"),
	UnitPriceBaseQuantity: types.StringPtr("unit_price_quantity"),
	UnitPriceBaseUnit:     types.StringPtr("unit_price_unit"),
	LowestPrice30d:        types.StringPtr("lowest_price_30d"),
	AnchorPrice:           types.StringPtr("anchor_price"),
	AnchorPriceAsOf:       types.StringPtr("anchor_price_date"),
}

// studenacFieldMappingAlt is the alternative field mapping (Croatian headers/uppercase)
var studenacFieldMappingAlt = xml.XmlFieldMapping{
	ExternalID:            types.StringPtr("Sifra"),
	Name:                  "Naziv",
	Description:           types.StringPtr("Opis"),
	Category:              types.StringPtr("Kategorija"),
	Subcategory:           types.StringPtr("Podkategorija"),
	Brand:                 types.StringPtr("Marka"),
	Unit:                  types.StringPtr("Jedinica"),
	UnitQuantity:          types.StringPtr("Kolicina"),
	Price:                 "Cijena",
	DiscountPrice:         types.StringPtr("AkcijskaCijena"),
	DiscountStart:         types.StringPtr("PocetakAkcije"),
	DiscountEnd:           types.StringPtr("KrajAkcije"),
	Barcodes:              types.StringPtr("Barkod"),
	ImageURL:              types.StringPtr("Slika"),
	UnitPrice:             types.StringPtr("CijenaZaJedinicuMjere"),
	UnitPriceBaseQuantity: types.StringPtr("JedinicaMjereKolicina"),
	UnitPriceBaseUnit:     types.StringPtr("JedinicaMjereOznaka"),
	LowestPrice30d:        types.StringPtr("NajnizaCijena30Dana"),
	AnchorPrice:           types.StringPtr("SidrenaCijena"),
	AnchorPriceAsOf:       types.StringPtr("SidrenaCijenaDatum"),
}

// studenacStoreExtractor extracts store ID from Studenac XML items
// Tries: store_id, storeId, Store.Id
func studenacStoreExtractor(item map[string]interface{}) string {
	// Try store_id
	if storeID, ok := item["store_id"]; ok {
		return toString(storeID)
	}
	// Try storeId
	if storeID, ok := item["storeId"]; ok {
		return toString(storeID)
	}
	// Try Store.Id (nested)
	if store, ok := item["Store"].(map[string]interface{}); ok {
		if id, ok := store["Id"]; ok {
			return toString(id)
		}
	}
	return ""
}

// studenacStoreExtractorAlt extracts store ID using alternative field names
// Tries: StoreId, STORE_ID, Poslovnica.Id
func studenacStoreExtractorAlt(item map[string]interface{}) string {
	// Try StoreId
	if storeID, ok := item["StoreId"]; ok {
		return toString(storeID)
	}
	// Try STORE_ID
	if storeID, ok := item["STORE_ID"]; ok {
		return toString(storeID)
	}
	// Try Poslovnica.Id (nested)
	if poslovnica, ok := item["Poslovnica"].(map[string]interface{}); ok {
		if id, ok := poslovnica["Id"]; ok {
			return toString(id)
		}
	}
	return ""
}

// toString converts interface to string
func toString(v interface{}) string {
	if v == nil {
		return ""
	}
	switch val := v.(type) {
	case string:
		return val
	case float64:
		return fmt.Sprintf("%.0f", val)
	case int:
		return fmt.Sprintf("%d", val)
	case int64:
		return fmt.Sprintf("%d", val)
	default:
		return fmt.Sprintf("%v", v)
	}
}

// StudenacAdapter is the chain adapter for Studenac retail chain
type StudenacAdapter struct {
	*base.BaseXmlAdapter
	discoveryDate string // Date to filter discovery (YYYY-MM-DD format)
}

// NewStudenacAdapter creates a new Studenac adapter
func NewStudenacAdapter() (*StudenacAdapter, error) {
	chainConfig := config.ChainConfigs[config.ChainStudenac]

	// Set up primary mapping with store extractor
	primaryMapping := studenacFieldMapping
	primaryMapping.NameExtractor = nil // Use path-based extraction for name
	primaryMapping.PriceExtractor = nil

	// Set up alternative mapping with store extractor
	altMapping := studenacFieldMappingAlt
	altMapping.NameExtractor = nil
	altMapping.PriceExtractor = nil

	adapterConfig := base.XmlAdapterConfig{
		BaseAdapterConfig: base.BaseAdapterConfig{
			Slug:           string(config.ChainStudenac),
			Name:           chainConfig.Name,
			SupportedTypes: []types.FileType{types.FileTypeXML},
			ChainConfig:    chainConfig,
			FilenamePrefixPatterns: []string{
				`(?i)^Studenac[_-]?`,
				`(?i)^cjenik[_-]?`,
				`(?i)^SUPERMARKET[_-]?`,
			},
		},
		FieldMapping:            primaryMapping,
		AlternativeFieldMapping: &altMapping,
		DefaultItemsPath:        "products.product",
		ItemPaths: []string{
			"products.product",
			"Products.Product",
			"Proizvodi.Proizvod",
			"proizvodi.proizvod",
			"items.item",
			"Items.Item",
		},
	}

	baseAdapter, err := base.NewBaseXmlAdapter(adapterConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create base XML adapter: %w", err)
	}

	return &StudenacAdapter{
		BaseXmlAdapter: baseAdapter,
	}, nil
}

// SetDiscoveryDate sets the date to filter discovery results
// Date should be in YYYY-MM-DD format
func (a *StudenacAdapter) SetDiscoveryDate(date string) {
	a.discoveryDate = date
}

// Discover discovers available XML price files from Studenac portal
// If SetDiscoveryDate was called, filters results by that date
func (a *StudenacAdapter) Discover(targetDate string) ([]types.DiscoveredFile, error) {
	discoveredFiles := make([]types.DiscoveredFile, 0)
	seenURLs := make(map[string]bool)

	// Use SetDiscoveryDate value if targetDate not provided
	filterDate := targetDate
	if filterDate == "" && a.discoveryDate != "" {
		filterDate = a.discoveryDate
	}

	fmt.Printf("[DEBUG] Fetching Studenac portal: %s\n", a.BaseURL())

	resp, err := a.HTTPClient().Get(a.BaseURL())
	if err != nil {
		return nil, fmt.Errorf("failed to fetch portal: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	body := string(bodyBytes)

	// Extract XML file links
	xmlPattern := regexp.MustCompile(`href=["']([^"']*\.xml(?:\?[^"']*)?)["']`)
	matches := xmlPattern.FindAllStringSubmatch(body, -1)

	for _, match := range matches {
		if len(match) < 2 {
			continue
		}

		href := match[1]
		fileURL := a.resolveURL(href)

		// Skip duplicates
		if seenURLs[fileURL] {
			continue
		}
		seenURLs[fileURL] = true

		// Extract filename from URL
		filename := a.extractFilenameFromURL(fileURL)
		fileDate := a.extractDateFromFilename(filename)

		// Filter by date if filterDate is set
		if filterDate != "" && fileDate != "" && fileDate != filterDate {
			continue
		}

		var lastModified *time.Time
		if fileDate != "" {
			if t, err := time.Parse("2006-01-02", fileDate); err == nil {
				lastModified = &t
			}
		}

		discoveredFiles = append(discoveredFiles, types.DiscoveredFile{
			URL:          fileURL,
			Filename:     filename,
			Type:         types.FileTypeXML,
			Size:         nil,
			LastModified: lastModified,
			Metadata: map[string]string{
				"source":       "studenac_portal",
				"discoveredAt": time.Now().Format(time.RFC3339),
				"portalDate":   fileDate,
			},
		})
	}

	return discoveredFiles, nil
}

// resolveURL resolves a potentially relative URL against the base URL
func (a *StudenacAdapter) resolveURL(href string) string {
	if strings.HasPrefix(href, "http://") || strings.HasPrefix(href, "https://") {
		return href
	}

	baseURL := a.BaseURL()
	parsed, err := url.Parse(baseURL)
	if err != nil {
		return baseURL + "/" + href
	}

	if strings.HasPrefix(href, "/") {
		// Absolute path
		return parsed.Scheme + "://" + parsed.Host + href
	}

	// Relative URL
	return baseURL + "/" + href
}

// extractFilenameFromURL extracts filename from URL
func (a *StudenacAdapter) extractFilenameFromURL(fileURL string) string {
	parsed, err := url.Parse(fileURL)
	if err != nil {
		return fileURL
	}

	path := parsed.Path
	parts := strings.Split(path, "/")
	if len(parts) > 0 {
		return parts[len(parts)-1]
	}

	return fileURL
}

// extractDateFromFilename extracts date from Studenac filename
// Tries YYYY-MM-DD and DD-MM-YYYY patterns
// Example: SUPERMARKET-Bijela_uvala_5_FUNTANA-T598-229-2026-12-29-07-00-14-559375.xml
func (a *StudenacAdapter) extractDateFromFilename(filename string) string {
	// Try YYYY-MM-DD pattern
	isoPattern := regexp.MustCompile(`(\d{4})-(\d{2})-(\d{2})`)
	if match := isoPattern.FindStringSubmatch(filename); len(match) == 4 {
		return fmt.Sprintf("%s-%s-%s", match[1], match[2], match[3])
	}

	// Try DD-MM-YYYY pattern
	euPattern := regexp.MustCompile(`(\d{2})-(\d{2})-(\d{4})`)
	if match := euPattern.FindStringSubmatch(filename); len(match) == 4 {
		return fmt.Sprintf("%s-%s-%s", match[3], match[2], match[1])
	}

	return ""
}

// ExtractStoreIdentifierFromFilename extracts store identifier from Studenac filename
// Pattern: {TYPE}-{LOCATION}-T{CODE}-{DATE...}.xml
func (a *StudenacAdapter) ExtractStoreIdentifierFromFilename(filename string) string {
	// Try to extract T-code (e.g., T598)
	tCodePattern := regexp.MustCompile(`-T(\d+)-`)
	if match := tCodePattern.FindStringSubmatch(filename); len(match) >= 2 {
		return match[1]
	}

	// Fallback to base class method
	dummyFile := types.DiscoveredFile{Filename: filename}
	if id := a.ExtractStoreIdentifier(dummyFile); id != nil {
		return id.Value
	}

	return ""
}

// ExtractStoreMetadata extracts store metadata from Studenac filename for auto-registration
// Pattern: {TYPE}-{LOCATION}-T{CODE}-{DATE...}.xml
// Example: SUPERMARKET-Bijela_uvala_5_FUNTANA-T598-229-2026-12-29-07-00-14-559375.xml
func (a *StudenacAdapter) ExtractStoreMetadata(file types.DiscoveredFile) *types.StoreMetadata {
	// Extract type and location from pattern: {TYPE}-{LOCATION}-T{CODE}-
	pattern := regexp.MustCompile(`^([A-Z]+)-(.+?)-T\d+-`)
	match := pattern.FindStringSubmatch(file.Filename)

	if len(match) < 3 {
		// Fall back to default behavior
		storeID := a.ExtractStoreIdentifierFromFilename(file.Filename)
		return &types.StoreMetadata{
			Name: fmt.Sprintf("%s %s", a.Name(), storeID),
		}
	}

	storeType := match[1]                           // "SUPERMARKET"
	locationRaw := match[2]                         // "Bijela_uvala_5_FUNTANA"
	location := strings.ReplaceAll(locationRaw, "_", " ") // "Bijela uvala 5 FUNTANA"

	// Try to separate address and city (city is often last word in ALL CAPS)
	words := strings.Split(location, " ")
	lastWord := words[len(words)-1]

	if isAllUpper(lastWord) && len(words) > 1 {
		// Last word is city (all caps like FUNTANA)
		city := lastWord
		address := strings.Join(words[:len(words)-1], " ")
		return &types.StoreMetadata{
			Name:      fmt.Sprintf("Studenac %s", titleCase(city)),
			Address:   titleCase(address),
			City:      titleCase(city),
			StoreType: storeType,
		}
	}

	return &types.StoreMetadata{
		Name:      fmt.Sprintf("Studenac %s", titleCase(location)),
		Address:   titleCase(location),
		StoreType: storeType,
	}
}

// Parse overrides the base Parse to inject store ID extractor
func (a *StudenacAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	// First try with primary mapping
	result, err := a.BaseXmlAdapter.Parse(content, filename, options)
	if err != nil {
		return nil, err
	}

	// Post-process to extract store IDs from items if not already set
	storeID := a.ExtractStoreIdentifierFromFilename(filename)
	for i := range result.Rows {
		if result.Rows[i].StoreIdentifier == "" {
			result.Rows[i].StoreIdentifier = storeID
		}
	}

	return result, nil
}

// isAllUpper checks if a string is all uppercase
func isAllUpper(s string) bool {
	for _, r := range s {
		if r >= 'a' && r <= 'z' {
			return false
		}
	}
	return true
}

// titleCase converts string to title case
func titleCase(s string) string {
	words := strings.Split(strings.ToLower(s), " ")
	for i, word := range words {
		if len(word) > 0 {
			words[i] = strings.ToUpper(string(word[0])) + word[1:]
		}
	}
	return strings.Join(words, " ")
}
